Modules.md

```diff
@@ -1,58 +1,95 @@
-
-While WebAssembly modules are designed to interoperate with ES6 modules
-in a Web environment (more details [below](Modules.md#integration-with-es6-modules)),
-WebAssembly modules are defined independently of JavaScript and do not require
-the host environment to include a JavaScript VM.
-
-## Imports and Exports
-
-A module defines a set of functions in its
-[code section](Modules.md#code-section) and can declare and name a subset of
-these functions to be **exports**. The meaning of exports (how and when they are
-called) is defined by the host environment. For example, a minimal shell
-environment might only probe for and call the start function defined by the start
-node of the module when given a module to execute.  Exports are exported by name,
-where the name is an arbitrary byte string of a given length. The host may need
-to mangle these names.
-
-A module can declare a set of **imports**. An import is a tuple containing a
-module name, the name of an exported function to import from the named module,
-and the signature to use for that import within the importing module. Within a
-module, the import can be [directly called](AstSemantics.md#calls) like a
-function (according to the signature of the import). When the imported
-module is also WebAssembly, it would be an error if the signature of the import
-doesn't match the signature of the export.
-
-The WebAssembly spec does not define how imports are interpreted:
+* the code of the instantiated module; and
+* a set of [index spaces](#index-spaces) of linear memories, tables and
+  functions that were imported or defined by the module.
+
+## Index Spaces
+
+Any number of functions, memories and tables can be both imported and
+defined in a given module. Imports and definitions are merged into index
+spaces (one for each of memories, tables and functions) which allows uniform
+treatment of imports and definitions thereafter (via index).
+
+Specifically, the index space of functions is defined by first giving each of
+the M function imports an index \[0, M\) (ordered by their sequence in the
+imports section) and then giving each of the N function definitions an index 
+\[M, M+N\) (ordered by their sequence in the module). The memory and table
+index spaces are defined in the same way.
+
+Thus, any use of a *function index* (e.g., a [`call`](AstSemantics.md#calls))
+can refer to either a function import or definition and similarly for
+and *table index* or *memory index*.
+
+## Imports
+
+A module can declare a sequence of **imports** which are provided, at
+instantiation time, by the host environment. There are several kinds of imports:
+* **function imports**, which can be called inside the module by the
+  [`call`](AstSemantics.md#calls) operator;
+* **linear memory imports**, which can be accessed inside the module by the
+  [memory operators](AstSemantics.md#linear-memory); and
+* **table imports**, which can be accessed inside the module by 
```

Comment:
Yes, high-performance access to host-provided "memory-like things" is a definitely a use case here ((`Shared`)`ArrayBuffer`, something from a `<canvas>`, maybe a server wasm environment can have an efficient zero-copy network API, ...)